`X509_STORE_CTX_set_default()`: improve error handling and reporting, also in its use

DDvO opened the pull request:
So far, error result of `X509_STORE_CTX_set_default()` was not checked in `pk7_smime.c` and the error code was misleading in in `x509_vfy.c`.
This PR also adds error information which purpose name could not be found.

As the bugs fixed are minor and `ERR_raise_data()` is not available in 1.1.1., I do not suggest backporting to 1.1.1.

kaduk commented on crypto/pkcs7/pk7_smime.c:
```diff
@@ -280,7 +280,8 @@ int PKCS7_verify(PKCS7 *p7, STACK_OF(X509) *certs, X509_STORE *store,
                 }
-                X509_STORE_CTX_set_default(cert_ctx, "smime_sign");
+                if (!X509_STORE_CTX_set_default(cert_ctx, "smime_sign"))
+                    goto err;
```
Why do we not add an error to the stack here?

DDvO replied:
We do not need because the error stack entry is already added by the function being called: `X509_STORE_CTX_set_default()`
I generally prefer this style because the called function knows much better what went wrong and why.

kaduk commented on crypto/x509/x509_vfy.c:
```diff
@@ -2430,15 +2430,11 @@ int X509_STORE_CTX_init(X509_STORE_CTX *ctx, X509_STORE *store, X509 *x509,
     else
         ctx->param->inh_flags |= X509_VP_FLAG_DEFAULT | X509_VP_FLAG_ONCE;
+    if (ret == 0)
+        goto err;
```
Why do we not add an error to the stack here?

DDvO replied:
Basically for the same reason as mentioned above: the called function, `X509_VERIFY_PARAM_inherit()` in this case, knows much better what went wrong and why.
And `ERR_R_MALLOC_FAILURE`, given here before, may be even wrong/misleading at this point.
If there is actually an malloc failure, this should be flagged by the respective allocation function itself.

kaduk reviewed (commented):
The core of the changes looks okay, but I am not sure whether we want to try hard to make functions consistently add (or not add) to the error stack on failure.

DDvO commented:
Ping @openssl/committers for 2nd approval.

kaduk commented:
> We do not need because the error stack entry is already added by the function being called: X509_STORE_CTX_set_default()
I generally prefer this style because the called function knows much better what went wrong and why.

I would have a lot more sympathy for this if `X509_STORE_CTX_set_default()` actually did consistently add an entry to the error stack.  (It seems that it does ... for some error paths but not all.)

As an exercise, I traced through the potential error returns from `X509_STORE_CTX_set_default()`, indicating (at corresponding indentation level) the called function whose failure might trickle up to a zero return from `X509_STORE_CTX_set_default()` itself.  The following is incomplete, as I did not descend into all the stack and OBJ routines.  It also appears that, while custom mallocs are passed file and line information so as to be able to provide some useful error information, the default `CRYPTO_malloc()` does not actually add to the error stack.

```
X509_STORE_CTX_set_default:
  X509_VERIFY_PARAM_lookup:
    sk_X509_VERIFY_PARAM_value:
    OBJ_bsearch_table:
  X509_VERIFY_PARAM_inherit:
    X509_VERIFY_PARAM_set1_policies:
      [arg is NULL, does not add error]
      sk_ASN1_OBJECT_new_null():
      OBJ_dup:
      sk_ASN1_OBJECT_push:
    sk_OPENSSL_STRING_deep_copy:
      [calls ERR_raise(., ERR_R_MALLOC_FAILURE) on error]
    X509_VERIFY_PARAM_set1_email:
      int_x509_param_set1:
        OPENSSL_malloc:
    X509_VERIFY_PARAM_set1_ip:
      [iplen not in (0, 4, 16), does not add error]
      int_x509_param_set1:
        OPENSSL_malloc:
```
As noted, `X509_VERIFY_PARAM_set1_policies()` and `X509_VERIFY_PARAM_set1_ip()` both have sites that return an error directly based on processing their input arguments (not due to calling a subroutine) and do not add to the error stack.  So I conclude that `X509_STORE_CTX_set_default()` cannot be relied upon to do this.  If you believe that in the platonic ideal state, we should be able to rely upon it to do so, then please open a separate issue (or add more commits to this PR fixing it, I suppose), and I will be willing to approve (though see below).

> And ERR_R_MALLOC_FAILURE, which was given here before, may be even wrong/misleading at this point.
If there is actually an malloc failure, this should be flagged by the respective allocation function itself.

Note that `ERR_raise()` includes the file/line information from where it was called.  Whether the file/line information from higher or lower in the call stack will be most useful in debugging will depend on the specific failure scenario, so I don't think we can make a uniform statement about one always being better or worse than the other.  That said, even if we assume for sake of argument that we will consistently add to the error stack directly above the `OPENSSL_malloc()` failure, the implementation of `X509_STORE_CTX_set_default()` does not currently do so.

DDvO commented:
Thanks @kaduk for having a closer look and for your further comments.

>> We do not need because the error stack entry is already added by the function being called: `X509_STORE_CTX_set_default()`
>> I generally prefer this style because the called function knows much better what went wrong and why.
>
> I would have a lot more sympathy for this if `X509_STORE_CTX_set_default()` actually did consistently add an entry to the error stack. (It seems that it does ... for some error paths but not all.)

Yeah, though this is not the fault of `X509_STORE_CTX_set_default()` but of some VPM and stack functions it uses.

> As an exercise, I traced through the potential error returns from `X509_STORE_CTX_set_default()`, indicating (at corresponding indentation level) the called function whose failure might trickle up to a zero return from `X509_STORE_CTX_set_default()` itself. The following is incomplete, as I did not descend into all the stack and OBJ routines. It also appears that, while custom mallocs are passed file and line information so as to be able to provide some useful error information, the default `CRYPTO_malloc()` does not actually add to the error stack.
> 
> ```
> X509_STORE_CTX_set_default:
>   X509_VERIFY_PARAM_lookup:
>     sk_X509_VERIFY_PARAM_value:
>     OBJ_bsearch_table:
>   X509_VERIFY_PARAM_inherit:
>     X509_VERIFY_PARAM_set1_policies:
>       [arg is NULL, does not add error]
>       sk_ASN1_OBJECT_new_null():
>       OBJ_dup:
>       sk_ASN1_OBJECT_push:
>     sk_OPENSSL_STRING_deep_copy:
>       [calls ERR_raise(., ERR_R_MALLOC_FAILURE) on error]
>     X509_VERIFY_PARAM_set1_email:
>       int_x509_param_set1:
>         OPENSSL_malloc:
>     X509_VERIFY_PARAM_set1_ip:
>       [iplen not in (0, 4, 16), does not add error]
>       int_x509_param_set1:
>         OPENSSL_malloc:
> ```
> 
> As noted, `X509_VERIFY_PARAM_set1_policies()` and `X509_VERIFY_PARAM_set1_ip()` both have sites that return an error directly based on processing their input arguments (not due to calling a subroutine) and do not add to the error stack. So I conclude that `X509_STORE_CTX_set_default()` cannot be relied upon to do this. If you believe that in the platonic ideal state, we should be able to rely upon it to do so, then please open a separate issue (or add more commits to this PR fixing it, I suppose), and I will be willing to approve (though see below).

I fixed these and related issues in `x509_vpm.c` in a separate commit.
 
> > And ERR_R_MALLOC_FAILURE, which was given here before, may be even wrong/misleading at this point.
> > If there is actually an malloc failure, this should be flagged by the respective allocation function itself.
> 
> Note that `ERR_raise()` includes the file/line information from where it was called. Whether the file/line information from higher or lower in the call stack will be most useful in debugging will depend on the specific failure scenario, so I don't think we can make a uniform statement about one always being better or worse than the other. That said, even if we assume for sake of argument that we will consistently add to the error stack directly above the `OPENSSL_malloc()` failure, the implementation of `X509_STORE_CTX_set_default()` does not currently do so.

This is an instance of the general malloc error reporting problem I tried to tackle in https://github.com/openssl/openssl/pull/14833 but that got stuck.
So as a more direct solution I fixed the malloc reportin and related issues at the levelo of `stack.c` in a separate commit.

Doing so, I made use of `CRYPTO_R_TOO_MANY_RECORDS` at some places, which confirmed my suspicion that `ERR_R_MALLOC_FAILURE` is occasionally misleading.
Adding `ERR_R_PASSED_INVALID_ARGUMENT` where it made sense to me made some tests fail, which revealed some sloppiness in using the stack API in `crypto/conf/conf_def.c`, `ssl/ssl_ciph.c`, `ssl/statem/statem_srvr.c`, and mostly in `test/helpers/ssltestlib.c`. So I fixed that as well.

DDvO commented:
For some reason, a couple of CI tests fail when build with `-no-tls1_3`,
which I can reproduce locally like this:
```
make  test TESTS=test_ssl_new SSL_TESTS=12-ct.cnf
```
but when I try to reproduce this in a shell or Eclipse (such that I can use `gdb`) with the current directory being `test-runs/test_ssl_new` via
```
../../test/ssl_test 12-ct.cnf.none none
```
I get a different type of failure:
```
# ERROR: (int) 'NCONF_load(conf, test_get_argument(0), NULL) > 0' failed @ test/ssl_test.c:575
# [0] compared to [0]
# ERROR:  @ test/ssl_test.c:578
# usage: ssl_test conf_file module_name [module_conf_file]
```
which for some weird reason is avoided by the above `make test` invocation by
```
TOP=../.. ../../util/wrap.pl /usr/bin/perl -I ../../util/perl ../../test/generate_ssl_tests.pl ../../test/ssl-tests/12-ct.cnf.in none > 12-ct.cnf.none
```

Can anyone tell me how to get
```
../../test/ssl_test 12-ct.cnf.none none
```
started properly, without the ERROR given above?

kaduk commented:
IIRC the test suite has also set a few environment variables at this point, though I would have to look up which ones.

kaduk commented:
> IIRC the test suite has also set a few environment variables at this point, though I would have to look up which ones.

in `test-runs/test_ssl_new`, running `../../util/wrap.pl ../../test/ssl_test 12-ct.cnf.default default` produces the same "NCONF_load" error you see.  Running `CTLOG_FILE=../../test/ct/log_list.cnf TEST_CERTS_DIR=<path to repo_root>/test/certs SRCTOP=<path to repo root> BLDTOP=<path to repo root> ../../util/wrap.pl ../../test/ssl_test 12-ct.cnf.default default` succeeds.  It seems that `CTLOG_FILE` is implicitly documented in `test/README.ssltest.md` in the examples, but not why it is specifically needed.

DDvO commented:
The unexpected implicit dependencies on `CTLOG_FILE` and `TEST_CERTS_DIR` should be documented
at least in `test/recipes/80-test_ssl_new.t` - I'll do this here (in a separate commit, of course).

DDvO commented:
@kaduk I've just returned from vacation and with the help of your above response I've been able to reproduce and fix the remaining CI failure.
So your above review comments are now addressed in full.

openssl-machine commented:
This PR is in a state where it requires action by @openssl/committers but the last update was 30 days ago

DDvO commented:
Ping for 2nd review

slontis commented on crypto/x509/x509_vfy.c:
```diff
@@ -2430,15 +2430,11 @@ int X509_STORE_CTX_init(X509_STORE_CTX *ctx, X509_STORE *store, X509 *x509,
         ret = X509_VERIFY_PARAM_inherit(ctx->param, store->param);
     else
         ctx->param->inh_flags |= X509_VP_FLAG_DEFAULT | X509_VP_FLAG_ONCE;
+    if (ret == 0)
```
get rid of the ret variable if it is just used here and then thrown away...
i.e.
````
if () {
   if (X509-VERIFY...()== 0)
     goto err;
 } else {
   ctx->param->,,,,,, |= ...;
 }
}
````

DDvO replied:
Good spotting.
Done (though this is 'just' refactoring), in a simplified form.

slontis commented on ssl/ssl_ciph.c:
```diff
@@ -532,7 +532,8 @@ int ssl_cipher_get_evp(SSL_CTX *ctx, const SSL_SESSION *s,
             i = sk_SSL_COMP_find(ssl_comp_methods, &ctmp);
-            *comp = sk_SSL_COMP_value(ssl_comp_methods, i);
+            if (i >= 0)
+                *comp = sk_SSL_COMP_value(ssl_comp_methods, i);
```
This almost seems like an error state

DDvO replied:
You mean, the case `i < 0` now excluded by the new guard?
Could well be, but I have no insight in what exactly this function is supposed to do in this (corner) case. If there is anything to fix here, I'd say it goes beyond the scope of this PR.

DDvO commented:
@t8m thanks for your swift original approval.
Can you please re-approve?

openssl-machine commented:
24 hours has passed since 'approval: done' was set, but as this PR has been updated in that time the label 'approval: ready to merge' is not being automatically set.  Please review the updates and set the label manually.

DDvO commented:
Merged to master and 3.0 - thanks @t8m and @slontis
